### PR TITLE
Splitted rep_gen.py, rep_gen_test.py and custom_train.py into new branch

### PR DESCRIPTION
**Pull request by airbagy**

### CODE REVIEW

File: rep_gen_test.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+        # Assert that:
+        #   1. Output files are directed to out_dir
+        #   2. Output files have the correct names
+        #   3. Output files have the correct format
+        #   4. Output files have the correct value
+        #   5. Argument char_as_label work as expected
+
+        # Write codepoints and embeddings to file
+        rg = RepresentationGenerator(out_dir=self.tmp_dir)
```
Comment (FrankYFTang) [inline rep_gen_test.py:178]: please add code to assure the file self.tmp_dir, filename + '_vec.tsv' and self.tmp_dir, filename + '_meta.tsv' does not exist before you call write_embeddings_from_list

File: rep_gen_test.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        # Read newly generated meta file
+        with open(os.path.join(self.tmp_dir, filename + '_meta.tsv'),'r') \
+            as f_in:
+            output_chars = [label.strip() for label in f_in.readlines()]
+        # Check vec file content
+        self.assertTrue(np.array_equal(embeddings, output_embs))
+        # Check meta file cntent
+        self.assertFalse(codepoints == output_chars)
+        self.assertTrue(output_chars == ['一', '丁'])
+
```
Comment (FrankYFTang) [inline rep_gen_test.py:210]: Please add add code to remove the file self.tmp_dir, filename + '_vec.tsv' and self.tmp_dir, filename + '_meta.tsv' at this point.

File: rep_gen_test.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        # Write codepoints and embeddings to file
```
Comment (FrankYFTang) [inline rep_gen_test.py:211]: Please split the code after this to a separated test function such as
def test_write_embeddings_from_list_not_as_label(self)
and make sure you remove / clean up the generated test file in your test.

File: rep_gen_test.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        rg = RepresentationGenerator()
+        # Get mocked filename dataset
+        mock_img = MagicMock()
+        mock_label = MagicMock()
+        mock_ds = [(mock_img, mock_label)]
+
+        # Mock function get_filename_dataset.
+        with patch.object(rg._dataset_builder, 'get_filename_dataset',
+                          return_value=mock_ds) as get_fd:
+            codepoints, embeddings = rg.get_embeddings(img_dir=self.tmp_dir)
```
Comment (FrankYFTang) [inline rep_gen_test.py:100]: you should assert the codepoints, embeddings values returned by  get_embeddings are correct based on what the mock object would return. Take a look of side_effect  and return_value in https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.side_effect

File: custom_train.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+        # Build optimizer
+        optimizer = self._OPTIMIZER_MAP[self._TPL_OPTIMIZER](self._TPL_LR_VALUE)
+
+        # Load initial weights from self._TPL_INIT_DIR
+        init_dir = self._TPL_INIT_DIR
+        latest = tf.train.latest_checkpoint(init_dir)
+        model.load_weights(latest)
+
+        # Freeze specified variables
```
Comment (FrankYFTang) [inline custom_train.py:179]: define a private function freeze_variables and put line 158-166 into that private function and call from here.

File: custom_train.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+        for layer in freeze_layers:
+            print('Freezing layer {}.'.format(layer.name))
+            layer.trainable = False
+
+        # Create loss function
+        loss = self._LOSS_MAP['TripletSemiHard'](self._TPL_MARGIN)
+        model.compile(optimizer=optimizer, loss=loss)
+
+        # Train triplet model
+        for i in range(self._TPL_CYCLES):
```
Comment (FrankYFTang) [inline custom_train.py:192]: I don't get what you are doing here. You have this loop and repeating the task, would get_train_dataset return different result each time? or we are feeding exactly the samething into model.fit() self._TPL_CYCLES times?

Comment (airbagy) [reply custom_train.py:192]: Yes. get_train_dataset will return a different dataset each time. All datasets would have only 40 (filter_size) randomly selected classes. This is the change I made so that triplet loss would converge. Will add more comments!

**Review by FrankYFTang — APPROVED** [review]
LGTM

**Comment by FrankYFTang** [comment]
feel free to merge this one AS IS.
